File: bindings/nodejs/test/fingerprint.test.js
'use strict';

// Tokenizer-fingerprint tests for the Node.js binding. Mirror the
// rust/dotnet/java fingerprint tests: 32-byte length + determinism,
// plus a cross-binding golden value for the byte_id pipeline.

const test = require('node:test');
const assert = require('node:assert/strict');

const ztok = require('..');

// Golden fingerprint for the default byte_id pipeline. Computed directly
// from libztok's ztok_fingerprint and shared across the
// rust/dotnet/java/python/ruby/go bindings to confirm agreement.
const GOLDEN_BYTE_ID =
    '201ecf86554b5a970471e0189d7e78dc2c3df24519f7d5ebb0caacc86701e77c';

test('fingerprint is 32 bytes and non-zero', () => {
    const pipe = ztok.Pipeline.byteId();
    try {
        const fp = pipe.fingerprint();
        assert.ok(fp instanceof ztok.Fingerprint);
        assert.equal(fp.bytes.length, 32);
        assert.ok(fp.bytes.some((b) => b !== 0));
        assert.equal(fp.hex().length, 64);
    } finally {
        pipe.close();
    }
});

test('fingerprint is deterministic for same config', () => {
    const a = ztok.Pipeline.byteId();
    const b = ztok.Pipeline.byteId();
    try {
        const fa = a.fingerprint();
        const fb = b.fingerprint();
        assert.equal(fa.hex(), fb.hex());
        assert.ok(fa.equals(fb));
    } finally {
        a.close();
        b.close();
    }
});

test('fingerprint matches cross-binding golden value', () => {
    const pipe = ztok.Pipeline.byteId();
    try {
        assert.equal(pipe.fingerprint().hex(), GOLDEN_BYTE_ID);
    } finally {
        pipe.close();
    }
});

File: bindings/go/fingerprint_test.go
package ztok

// Tokenizer-fingerprint tests over a byte_id pipeline. Mirror the
// rust/dotnet/java fingerprint tests: 32-byte length + determinism, plus
// a cross-binding golden value shared with the
// python/nodejs/ruby/rust/dotnet/java bindings.

import (
	"testing"
)

// goldenByteIDFingerprint is the fingerprint of the default byte_id
// pipeline. Computed directly from libztok's ztok_fingerprint and shared
// across all bindings to confirm cross-binding agreement.
const goldenByteIDFingerprint = "201ecf86554b5a970471e0189d7e78dc2c3df24519f7d5ebb0caacc86701e77c"

func TestFingerprintIs32Bytes(t *testing.T) {
	p, err := NewByteID(nil)
	if err != nil {
		t.Fatalf("NewByteID: %v", err)
	}
	defer p.Close()

	fp, err := p.Fingerprint()
	if err != nil {
		t.Fatalf("Fingerprint: %v", err)
	}
	if len(fp) != FingerprintSize {
		t.Fatalf("fingerprint length = %d, want %d", len(fp), FingerprintSize)
	}
	if len(fp.Hex()) != 64 {
		t.Fatalf("hex length = %d, want 64", len(fp.Hex()))
	}
	allZero := true
	for _, b := range fp {
		if b != 0 {
			allZero = false
			break
		}
	}
	if allZero {
		t.Fatal("fingerprint is all zeros")
	}
}

func TestFingerprintIsDeterministic(t *testing.T) {
	a, err := NewByteID(nil)
	if err != nil {
		t.Fatalf("NewByteID a: %v", err)
	}
	defer a.Close()
	b, err := NewByteID(nil)
	if err != nil {
		t.Fatalf("NewByteID b: %v", err)
	}
	defer b.Close()

	fa, err := a.Fingerprint()
	if err != nil {
		t.Fatalf("Fingerprint a: %v", err)
	}
	fb, err := b.Fingerprint()
	if err != nil {
		t.Fatalf("Fingerprint b: %v", err)
	}
	if fa != fb {
		t.Fatalf("fingerprints differ: %s vs %s", fa.Hex(), fb.Hex())
	}
}

func TestFingerprintGoldenByteID(t *testing.T) {
	p, err := NewByteID(nil)
	if err != nil {
		t.Fatalf("NewByteID: %v", err)
	}
	defer p.Close()

	fp, err := p.Fingerprint()
	if err != nil {
		t.Fatalf("Fingerprint: %v", err)
	}
	if got := fp.Hex(); got != goldenByteIDFingerprint {
		t.Fatalf("fingerprint = %s, want %s", got, goldenByteIDFingerprint)
	}
}
